File: deduce.ts
const text = await Deno.readTextFile("baza-wiedzy.txt");

const [initialRules, initialFacts] = text.split("fakty");

const facts = new Set(initialFacts.split(",").map((str) => str.trim()));

const rules = initialRules
  .split("\n")
  .filter(Boolean)
  .map((str) =>
    str
      .replaceAll(" ", "")
      .replace(/\d+\./g, "")
      .split("->")
      .map((str) => str.split(","))
  );

function deduce(hypothesis: string) {
  if (facts.has(hypothesis)) {
    return true;
  }

  let changed = true;
  while (changed) {
    changed = false;

    for (const [requirements, results] of rules) {
      if (results.every((r) => facts.has(r))) {
        continue;
      }

      if (requirements.every((r) => facts.has(r))) {
        changed = true;
        for (const result of results) {
          facts.add(result);
        }
      }
    }
  }

  return facts.has(hypothesis);
}

function deduceBackwards(hypothesis: string) {
  if (facts.has(hypothesis)) {
    return true;
  }

  for (const [requirements, results] of rules) {
    if (!results.includes(hypothesis)) {
      continue;
    }
    
    if (requirements.every((r) => deduceBackwards(r))) {
      for (const result of results) {
        facts.add(result);
      }

      return true;
    }
  }

  return false;
}

console.log(deduceBackwards("D"));
console.log(facts);

File: deduce.py
import re

text = open("baza-wiedzy-negacja.txt")

initials = re.split(r"fakty.*?\n", text.read(), flags=re.IGNORECASE)
text.close()

initialRules, initialFacts = tuple(initials)

rules = [tuple(s2.split(',') for s2 in re.sub(r'\d+\.', '', s.replace("~~", "").replace(" ", "")).split("->"))
         for s in initialRules.split('\n') if s]


def check(hypothesis, facts):
    return (hypothesis.replace('~', '') in facts) ^ hypothesis.startswith('~')


def deduce(hypothesis, facts):
    """Funkcja wykonuje wnioskowanie w przód."""
    if check(hypothesis, facts):
        return True

    changed = True
    while changed:
        changed = False

        for (requirements, results), i in zip(rules, range(len(rules))):

            if not all(check(requirement, facts) for requirement in requirements):
                if not all(any(all(check(requirement, facts) for requirement in requirements) for requirements, results in rules if result in results) for result in results):
                    for result in results:
                        if result in facts:
                            facts.remove(result)
                            changed = True
                continue

            if all(r in facts for r in results):
                continue

            changed = True
            for result in results:
                facts.add(result)

    return hypothesis in facts


def deduceRight(hypothesis, facts):
    """Funkcja wykonuje wnioskowanie w tył."""
    if hypothesis in facts:
        return True

    for requirements, results in rules:
        if hypothesis not in results:
            continue

        if any(not deduceRight(r.replace('~', ''), facts) ^ r.startswith('~') for r in requirements):
            continue

        for result in results:
            facts.add(result)

        return True

    return False


for hypothesis in sorted(set(r.replace('~', '') for r in sum(list(sum(list(r), []) for r in rules), []))):
    if deduceRight(hypothesis, set(s.strip() for s in initialFacts.split(','))):
        print(hypothesis, "jest spełnione")
    else:
        print(hypothesis, "nie jest spełnione")
